File: bindings/python/src/zyxdb/transaction.py
"""Pythonic transaction wrapper for ZYX database."""

from __future__ import annotations

from typing import Any

from zyxdb.result import Result


class Transaction:
    """Context manager wrapper around the core Transaction.

    Usage::

        with db.begin_transaction() as tx:
            tx.execute("CREATE (n:Person {name: $name})", name="Alice")
            # auto-commits on normal exit

    Auto-commits on normal exit, auto-rolls back on exception.
    """

    def __init__(self, core_tx: Any) -> None:
        self._tx = core_tx

    def execute(self, cypher: str, **params: Any) -> Result:
        """Execute a Cypher query within this transaction."""
        if not self._tx.is_active:
            raise RuntimeError("Cannot execute: transaction is not active")
        return Result(self._tx.execute(cypher, **params))

    def commit(self) -> None:
        """Commit the transaction."""
        if not self._tx.is_active:
            raise RuntimeError("Cannot commit: transaction is not active")
        self._tx.commit()

    def rollback(self) -> None:
        """Roll back the transaction."""
        if not self._tx.is_active:
            raise RuntimeError("Cannot rollback: transaction is not active")
        self._tx.rollback()

    @property
    def is_active(self) -> bool:
        """Whether the transaction is still active."""
        return self._tx.is_active

    @property
    def is_read_only(self) -> bool:
        """Whether this is a read-only transaction."""
        return self._tx.is_read_only

    def __enter__(self) -> Transaction:
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> bool:
        if self._tx.is_active:
            if exc_type is not None:
                self._tx.rollback()
            else:
                self._tx.commit()
        return False

    def __repr__(self) -> str:
        status = "active" if self.is_active else "closed"
        mode = "read-only" if self.is_read_only else "read-write"
        return f"Transaction({status}, {mode})"

File: bindings/nodejs/lib/transaction.js
'use strict';

const { Result } = require('./result');

/**
 * Transaction wrapper with async operations.
 *
 * @example
 * const tx = await db.beginTransaction();
 * try {
 *   await tx.execute('CREATE (n:Person {name: $name})', { name: 'Alice' });
 *   await tx.commit();
 * } catch (e) {
 *   await tx.rollback();
 *   throw e;
 * }
 */
class Transaction {
    /**
     * @param {Object} nativeTx - Native transaction object from addon.
     * @private
     */
    constructor(nativeTx) {
        this._tx = nativeTx;
        this._queue = Promise.resolve();
    }

    /**
     * Execute a Cypher query within this transaction.
     * @param {string} cypher - The Cypher query string.
     * @param {Object} [params] - Optional query parameters.
     * @returns {Promise<Result>}
     */
    execute(cypher, params) {
        if (!this._tx) {
            throw new Error('Transaction already closed');
        }
        const tx = this._tx;
        const operation = this._queue.then(async () => {
            const raw = await tx.execute(cypher, params || undefined);
            return new Result(raw);
        });
        this._queue = operation.then(() => undefined, () => undefined);
        return operation;
    }

    /**
     * Commit the transaction.
     * @returns {Promise<void>}
     */
    commit() {
        if (!this._tx) {
            throw new Error('Transaction already closed');
        }
        const tx = this._tx;
        this._tx = null;
        const operation = this._queue.then(() => tx.commit());
        this._queue = operation.then(() => undefined, () => undefined);
        return operation;
    }

    /**
     * Roll back the transaction.
     * @returns {Promise<void>}
     */
    rollback() {
        if (!this._tx) {
            throw new Error('Transaction already closed');
        }
        const tx = this._tx;
        this._tx = null;
        const operation = this._queue.then(() => tx.rollback());
        this._queue = operation.then(() => undefined, () => undefined);
        return operation;
    }

    /**
     * Whether the transaction is still active.
     * @type {boolean}
     */
    get isActive() {
        return this._tx ? this._tx.isActive : false;
    }

    /**
     * Whether this is a read-only transaction.
     * @type {boolean}
     */
    get isReadOnly() {
        return this._tx ? this._tx.isReadOnly : false;
    }
}

module.exports = { Transaction };
